File: Plugins/FirebaseGoodies/Source/FirebaseGoodies/Private/Database/FGDatabaseRef.cpp
// Copyright (c) 2022 Nineva Studios

#include "Database/FGDatabaseRef.h"

#include "Async/Async.h"
#include "Database/FGDataSnapshot.h"
#include "Database/FGMutableData.h"
#include "Database/FGTransactionHandler.h"
#include "Database/FGTransactionTask.h"
#include "FirebaseGoodiesLog.h"

#include "Interface/IDatabaseRef.h"

#if PLATFORM_ANDROID
#include "Android/DatabaseRefAndroid.h"
#endif

#if PLATFORM_IOS
#include "IOS/DatabaseRefIOS.h"
#endif

#if (PLATFORM_WINDOWS || PLATFORM_MAC) && FG_ENABLE_EDITOR_SUPPORT
#include "Desktop/DatabaseRefDesktop.h"
#endif


UFGDatabaseRef::UFGDatabaseRef()
{
	if (StaticClass()->GetDefaultObject() != this)
	{
#if PLATFORM_ANDROID
		DatabaseRefImpl = MakeShareable(new DatabaseRefAndroid());
#endif
#if PLATFORM_IOS
		DatabaseRefImpl = MakeShareable(new DatabaseRefIOS());
#endif
#if (PLATFORM_WINDOWS || PLATFORM_MAC) && FG_ENABLE_EDITOR_SUPPORT
		DatabaseRefImpl = MakeShareable(new DatabaseRefDesktop());
#endif
#if (PLATFORM_WINDOWS || PLATFORM_MAC) && !FG_ENABLE_EDITOR_SUPPORT
		DatabaseRefImpl = nullptr;
#endif
	}
}

void UFGDatabaseRef::InitPlatformData(const FString& DatabaseUrl, const FString& InitialPath)
{
	if(DatabaseRefImpl)
		DatabaseRefImpl->Init(DatabaseUrl, InitialPath);

	RefPath = InitialPath.Right(1) == TEXT("/") ? InitialPath.LeftChop(InitialPath.Len() - 1) : InitialPath;
}

UFGDatabaseRef* UFGDatabaseRef::MakeDatabaseRootRef()
{
	UFGDatabaseRef* dbRef = NewObject<UFGDatabaseRef>();
	dbRef->InitPlatformData("", "");
	return dbRef;
}

UFGDatabaseRef* UFGDatabaseRef::MakeDatabaseRefFromPath(const FString& Path)
{
	UFGDatabaseRef* dbRef = NewObject<UFGDatabaseRef>();
	dbRef->InitPlatformData("", Path);
	return dbRef;
}

UFGDatabaseRef* UFGDatabaseRef::MakeDatabaseReference(const FString& DatabaseUrl, const FString& Path)
{
	UFGDatabaseRef* dbRef = NewObject<UFGDatabaseRef>();
	dbRef->InitPlatformData(DatabaseUrl, Path);
	return dbRef;
}

UFGDatabaseRef* UFGDatabaseRef::Child(const FString& Path) const
{
	FString NewPath = RefPath / Path;
	return MakeDatabaseRefFromPath(NewPath);
}

UFGDatabaseRef* UFGDatabaseRef::Parent() const
{
	return MakeDatabaseRefFromPath(GetPathToParent());
}

UFGDatabaseRef* UFGDatabaseRef::Root() const
{
	return MakeDatabaseRootRef();
}

void UFGDatabaseRef::SetValue(const FFGValueVariant& Value, const FFGValueVariant& Priority)
{
	if (Priority.Type == EValueType::Array || Priority.Type == EValueType::Map || Priority.Type == EValueType::Bool)
	{
		UE_LOG(LogFirebaseGoodies, Error, TEXT("Priority cannot be a collection or boolean. Has to be a numerical or string value"));
		return;
	}

	if(DatabaseRefImpl)
		DatabaseRefImpl->SetValue(Value, Priority);
}

void UFGDatabaseRef::SetTimestamp()
{
	if(DatabaseRefImpl)
		DatabaseRefImpl->SetTimestamp();
}

void UFGDatabaseRef::SetPriority(const FFGValueVariant& Priority)
{
	if (Priority.Type == EValueType::Array || Priority.Type == EValueType::Map || Priority.Type == EValueType::Bool)
	{
		UE_LOG(LogFirebaseGoodies, Error, TEXT("Priority cannot be a collection or boolean. Has to be a numerical or string value"));
		return;
	}

	if(DatabaseRefImpl)
		DatabaseRefImpl->SetPriority(Priority);
}

FString UFGDatabaseRef::GetKey() const
{
	if(DatabaseRefImpl)
		return DatabaseRefImpl->GetKey();

	return FString();
}

void UFGDatabaseRef::GetValue(const FOnDataChangedDelegate& OnDataReceived, const FOnCancelledDelegate& OnCancelled)
{
	if(DatabaseRefImpl)
		DatabaseRefImpl->GetValue(OnDataReceived, OnCancelled);
}

UFGDatabaseRef* UFGDatabaseRef::Push()
{
	FString ChildString;

	if(DatabaseRefImpl)
	{
		ChildString = DatabaseRefImpl->Push();
	}

	return Child(ChildString);
}

void UFGDatabaseRef::RemoveValue()
{
	if(DatabaseRefImpl)
		DatabaseRefImpl->RemoveValue();
}

void UFGDatabaseRef::UpdateChildren(const TMap<FString, FFGValueVariant>& Update)
{
	if(DatabaseRefImpl)
		DatabaseRefImpl->UpdateChildren(Update);
}

void UFGDatabaseRef::AddValueListener(const FOnDataChangedDelegate& OnDataChanged, const FOnCancelledDelegate& OnCancelled)
{
	if(DatabaseRefImpl)
		DatabaseRefImpl->AddValueListener(OnDataChanged, OnCancelled);
}

void UFGDatabaseRef::RemoveValueListener()
{
	if(DatabaseRefImpl)
		DatabaseRefImpl->RemoveValueListener();
}

void UFGDatabaseRef::AddChildListener(const FOnChildEventDelegate& OnChildEvent, const FOnCancelledDelegate& OnCancelled)
{
	if(DatabaseRefImpl)
		DatabaseRefImpl->AddChildListener(OnChildEvent, OnCancelled);
}

void UFGDatabaseRef::RemoveChildListener()
{
	if(DatabaseRefImpl)
		DatabaseRefImpl->RemoveChildListener();
}

void UFGDatabaseRef::KeepSynced(bool Sync)
{
	if(DatabaseRefImpl)
		DatabaseRefImpl->KeepSynced(Sync);
}

UFGDatabaseRef* UFGDatabaseRef::OrderByChild(const FString& Path)
{
	if(DatabaseRefImpl)
		DatabaseRefImpl->OrderByChild(Path);

	return this;
}

UFGDatabaseRef* UFGDatabaseRef::OrderByKey()
{
	if(DatabaseRefImpl)
		DatabaseRefImpl->OrderByKey();

	return this;
}

UFGDatabaseRef* UFGDatabaseRef::OrderByPriority()
{
	if(DatabaseRefImpl)
		DatabaseRefImpl->OrderByPriority();

	return this;
}

UFGDatabaseRef* UFGDatabaseRef::OrderByValue()
{
	if(DatabaseRefImpl)
		DatabaseRefImpl->OrderByValue();

	return this;
}

UFGDatabaseRef* UFGDatabaseRef::EndAtFloat(float Value, const FString& Key)
{
	if(DatabaseRefImpl)
		DatabaseRefImpl->EndAtFloat(Value, Key);

	return this;
}

UFGDatabaseRef* UFGDatabaseRef::EndAtString(const FString& Value, const FString& Key)
{
	if(DatabaseRefImpl)
		DatabaseRefImpl->EndAtString(Value, Key);

	return this;
}

UFGDatabaseRef* UFGDatabaseRef::StartAtFloat(float Value, const FString& Key)
{
	if(DatabaseRefImpl)
		DatabaseRefImpl->StartAtFloat(Value, Key);

	return this;
}

UFGDatabaseRef* UFGDatabaseRef::StartAtString(const FString& Value, const FString& Key)
{
	if(DatabaseRefImpl)
		DatabaseRefImpl->StartAtString(Value, Key);

	return this;
}

UFGDatabaseRef* UFGDatabaseRef::EqualToFloat(float Value, const FString& Key)
{
	if(DatabaseRefImpl)
		DatabaseRefImpl->EqualToFloat(Value, Key);

	return this;
}

UFGDatabaseRef* UFGDatabaseRef::EqualToBool(bool Value, const FString& Key)
{
	if(DatabaseRefImpl)
		DatabaseRefImpl->EqualToBool(Value, Key);

	return this;
}

UFGDatabaseRef* UFGDatabaseRef::EqualToString(const FString& Value, const FString& Key)
{
	if(DatabaseRefImpl)
		DatabaseRefImpl->EqualToString(Value, Key);

	return this;
}

UFGDatabaseRef* UFGDatabaseRef::LimitToFirst(int Limit)
{
	if(DatabaseRefImpl)
		DatabaseRefImpl->LimitToFirst(Limit);

	return this;
}

UFGDatabaseRef* UFGDatabaseRef::LimitToLast(int Limit)
{
	if(DatabaseRefImpl)
		DatabaseRefImpl->LimitToLast(Limit);

	return this;
}

void UFGDatabaseRef::OnDisconnectSetValue(const FFGValueVariant& Value)
{
	if(DatabaseRefImpl)
		DatabaseRefImpl->OnDisconnectSetValue(Value);
}

void UFGDatabaseRef::OnDisconnectSetTimestamp()
{
	if(DatabaseRefImpl)
		DatabaseRefImpl->OnDisconnectSetTimestamp();
}

void UFGDatabaseRef::OnDisconnectRemoveValue()
{
	if(DatabaseRefImpl)
		DatabaseRefImpl->OnDisconnectRemoveValue();
}

void UFGDatabaseRef::OnDisconnectUpdateChildren(const TMap<FString, FFGValueVariant>& Update)
{
	if(DatabaseRefImpl)
		DatabaseRefImpl->OnDisconnectUpdateChildren(Update);
}

void UFGDatabaseRef::CancelDisconnectOperations()
{
	if(DatabaseRefImpl)
		DatabaseRefImpl->CancelDisconnectOperations();
}

void UFGDatabaseRef::BindTransactionTask(UFGTransactionTask* Task)
{
	if(DatabaseRefImpl)
		DatabaseRefImpl->BindTransactionTask(Task);
}

void UFGDatabaseRef::RunTransaction()
{
	if(DatabaseRefImpl)
		DatabaseRefImpl->RunTransaction();
}

FString UFGDatabaseRef::GetPathToParent() const
{
	if (RefPath.IsEmpty())
		return RefPath;

	int SlashPos = RefPath.FindLastCharByPredicate([](TCHAR C) { return C == TEXT('/'); });
	return RefPath.LeftChop(RefPath.Len() - SlashPos);
}

File: Plugins/FirebaseGoodies/Source/FirebaseGoodies/Private/Java/FGDatabaseRef.java
// Copyright (c) 2022 Nineva Studios

package com.ninevastudios.unrealfirebase;

import android.util.Log;

import com.google.firebase.database.ChildEventListener;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.MutableData;
import com.google.firebase.database.Query;
import com.google.firebase.database.ServerValue;
import com.google.firebase.database.Transaction;
import com.google.firebase.database.ValueEventListener;

import java.util.HashMap;

@SuppressWarnings("unused")
public class FGDatabaseRef {

	private static final int TRANSACTION_SUCCESS = 0;
	private static final int TRANSACTION_ABORT = 1;

	public static native int doTransactionCallback(long mCallbackAddr, FGMutableData data);

	public static native void onTransactionCompleteCallback(long mCallbackAddr, boolean isSuccess, String errorMessage);

	public static native void onQueryErrorCallback(String errorMessage);

	private DatabaseReference mDbRef;
	private Query mQuery = null;

	public FGDatabaseRef(String databaseUrl, String path) {
		FirebaseDatabase database;
		if (databaseUrl.isEmpty()) {
			database = FirebaseDatabase.getInstance();
		} else {
			database = FirebaseDatabase.getInstance(databaseUrl);
		}

		if (path.isEmpty()) {
			mDbRef = database.getReference();
		} else {
			mDbRef = database.getReference(path);
		}
	}

	public void SetValue(Object value, Object priority) {
		Log.d("FirebaseGoodies", "SetValue");
		if (mDbRef != null) {
			Log.d("FirebaseGoodies", "SetValue2");
			if (priority == null)
				mDbRef.setValue(value);
			else
				mDbRef.setValue(value, priority);
		}
	}

	public void SetTimestamp() {
		if (mDbRef != null) {
			mDbRef.setValue(ServerValue.TIMESTAMP);
		}
	}

	public void SetPriority(Object priority) {
		if (mDbRef != null) {
			mDbRef.setPriority(priority);
		}
	}

	public String Push() {
		if (mDbRef != null) {
			DatabaseReference newDbRef = mDbRef.push();
			return newDbRef.getKey();
		}

		return "";
	}

	public void RemoveValue() {
		if (mDbRef != null) {
			mDbRef.removeValue();
		}
	}

	public void UpdateChildren(HashMap<String, Object> update) {
		if (mDbRef != null) {
			mDbRef.updateChildren(update);
		}
	}

	private ValueEventListener mCurrentValueListener = null;
	private ChildEventListener mCurrentChildListener = null;

	public String GetKey() {
		if (mDbRef != null) {
			return mDbRef.getKey();
		}

		return "";
	}

	public void GetValue(long callbackAddr) {
		if (mDbRef != null) {
			Query q = mQuery != null ? mQuery : mDbRef;
			if (mCurrentValueListener != null)
				q.removeEventListener(mCurrentValueListener);
			q.addListenerForSingleValueEvent(new FGDatabaseValueEventListener(callbackAddr));
		}
	}

	public void AddValueListener(long callbackAddr) {
		if (mDbRef != null) {
			Query q = mQuery != null ? mQuery : mDbRef;
			if (mCurrentValueListener != null)
				q.removeEventListener(mCurrentValueListener);
			mCurrentValueListener = q.addValueEventListener(new FGDatabaseValueEventListener(callbackAddr));
		}
	}

	public void RemoveValueListener() {
		if (mDbRef != null && mCurrentValueListener != null) {
			Query q = mQuery != null ? mQuery : mDbRef;
			q.removeEventListener(mCurrentValueListener);
			mCurrentValueListener = null;
		}
	}

	public void AddChildListener(long callbackAddr) {
		if (mDbRef != null) {
			Query q = mQuery != null ? mQuery : mDbRef;
			if (mCurrentChildListener != null)
				q.removeEventListener(mCurrentChildListener);
			mCurrentChildListener = q.addChildEventListener(new FGDatabaseChildEventListener(callbackAddr));
		}
	}

	public void RemoveChildListener() {
		if (mDbRef != null && mCurrentChildListener != null) {
			Query q = mQuery != null ? mQuery : mDbRef;
			q.removeEventListener(mCurrentChildListener);
			mCurrentChildListener = null;
		}
	}

	public void KeepSynced(boolean sync) {
		if (mDbRef != null) {
			mDbRef.keepSynced(sync);
		}
	}

	public void OrderByChild(String path) {
		if (mDbRef == null) return;

		try {
			Query q = mQuery != null ? mQuery : mDbRef;
			mQuery = q.orderByChild(path);
		} catch (Exception e) {
			onQueryErrorCallback(e.getMessage());
		}
	}

	public void OrderByKey() {
		if (mDbRef == null) return;

		try {
			Query q = mQuery != null ? mQuery : mDbRef;
			mQuery = q.orderByKey();
		} catch (Exception e) {
			onQueryErrorCallback(e.getMessage());
		}
	}

	public void OrderByPriority() {
		if (mDbRef == null) return;

		try {
			Query q = mQuery != null ? mQuery : mDbRef;
			mQuery = q.orderByPriority();
		} catch (Exception e) {
			onQueryErrorCallback(e.getMessage());
		}
	}

	public void OrderByValue() {
		if (mDbRef == null) return;

		try {
			Query q = mQuery != null ? mQuery : mDbRef;
			mQuery = q.orderByValue();
		} catch (Exception e) {
			onQueryErrorCallback(e.getMessage());
		}
	}

	public void StartAtFloat(double value, String key) {
		if (mDbRef == null) return;

		try {
			Query q = mQuery != null ? mQuery : mDbRef;
			if (key.isEmpty()) {
				mQuery = q.startAt(value);
			} else {
				mQuery = q.startAt(value, key);
			}
		} catch (Exception e) {
			onQueryErrorCallback(e.getMessage());
		}
	}

	public void StartAtString(String value, String key) {
		if (mDbRef == null) return;

		try {
			Query q = mQuery != null ? mQuery : mDbRef;
			if (key.isEmpty()) {
				mQuery = q.startAt(value);
			} else {
				mQuery = q.startAt(value, key);
			}
		} catch (Exception e) {
			onQueryErrorCallback(e.getMessage());
		}
	}

	public void EndAtFloat(double value, String key) {
		if (mDbRef == null) return;

		try {
			Query q = mQuery != null ? mQuery : mDbRef;
			if (key.isEmpty()) {
				mQuery = q.endAt(value);
			} else {
				mQuery = q.endAt(value, key);
			}
		} catch (Exception e) {
			onQueryErrorCallback(e.getMessage());
		}
	}

	public void EndAtString(String value, String key) {
		if (mDbRef == null) return;

		try {
			Query q = mQuery != null ? mQuery : mDbRef;
			if (key.isEmpty()) {
				mQuery = q.endAt(value);
			} else {
				mQuery = q.endAt(value, key);
			}
		} catch (Exception e) {
			onQueryErrorCallback(e.getMessage());
		}
	}

	public void EqualToFloat(double value, String key) {
		if (mDbRef == null) return;

		try {
			Query q = mQuery != null ? mQuery : mDbRef;
			if (key.isEmpty()) {
				mQuery = q.equalTo(value);
			} else {
				mQuery = q.equalTo(value, key);
			}
		} catch (Exception e) {
			onQueryErrorCallback(e.getMessage());
		}
	}

	public void EqualToBool(boolean value, String key) {
		if (mDbRef == null) return;

		try {
			Query q = mQuery != null ? mQuery : mDbRef;
			if (key.isEmpty()) {
				mQuery = q.equalTo(value);
			} else {
				mQuery = q.equalTo(value, key);
			}
		} catch (Exception e) {
			onQueryErrorCallback(e.getMessage());
		}
	}

	public void EqualToString(String value, String key) {
		if (mDbRef == null) return;

		try {
			Query q = mQuery != null ? mQuery : mDbRef;
			if (key.isEmpty()) {
				mQuery = q.equalTo(value);
			} else {
				mQuery = q.equalTo(value, key);
			}
		} catch (Exception e) {
			onQueryErrorCallback(e.getMessage());
		}
	}

	public void LimitToFirst(int limit) {
		if (mDbRef == null) return;

		try {
			Query q = mQuery != null ? mQuery : mDbRef;
			mQuery = q.limitToFirst(limit);
		} catch (Exception e) {
			onQueryErrorCallback(e.getMessage());
		}
	}

	public void LimitToLast(int limit) {
		if (mDbRef == null) return;

		try {
			Query q = mQuery != null ? mQuery : mDbRef;
			mQuery = q.limitToLast(limit);
		} catch (Exception e) {
			onQueryErrorCallback(e.getMessage());
		}
	}

	public void OnDisconnectSetValue(Object value) {
		if (mDbRef != null) {
			mDbRef.onDisconnect().setValue(value);
		}
	}

	public void OnDisconnectSetTimestamp() {
		if (mDbRef != null) {
			mDbRef.onDisconnect().setValue(ServerValue.TIMESTAMP);
		}
	}

	public void OnDisconnectRemoveValue() {
		if (mDbRef != null) {
			mDbRef.onDisconnect().removeValue();
		}
	}

	public void OnDisconnectUpdateChildren(HashMap<String, Object> update) {
		if (mDbRef != null) {
			mDbRef.onDisconnect().updateChildren(update);
		}
	}

	public void CancelDisconnectOperations() {
		if (mDbRef != null) {
			mDbRef.onDisconnect().cancel();
		}
	}

	public void RunTransaction(final long transactionObj) {
		if (mDbRef != null) {
			mDbRef.runTransaction(new Transaction.Handler() {
				@Override
				public Transaction.Result doTransaction(MutableData mutableData) {
					// this function is called multiple times and sometimes the data in null
					// if it is return success, it will be called with proper data later
					if (mutableData.getValue() == null) return Transaction.success(mutableData);

					FGMutableData data = new FGMutableData(mutableData);
					int result = doTransactionCallback(transactionObj, data);
					return result == TRANSACTION_SUCCESS ? Transaction.success(mutableData) : Transaction.abort();
				}

				@Override
				public void onComplete(DatabaseError databaseError, boolean b, DataSnapshot dataSnapshot) {
					String errorMessage = databaseError != null ? databaseError.getMessage() : "";
					onTransactionCompleteCallback(transactionObj, b, errorMessage);
				}
			});
		}
	}
}
